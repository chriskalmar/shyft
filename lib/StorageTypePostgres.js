
import {
  StorageType,
  DataTypeID,
  DataTypeInteger,
  DataTypeBigInt,
  DataTypeFloat,
  DataTypeBoolean,
  DataTypeString,
  DataTypeJson,
  DataTypeTimestamp,
  DataTypeTimestampTz,
  DataTypeDate,
  sortDataByKeys,
  processCursors,
  processFilter,
  isMap,
  MUTATION_TYPE_CREATE,
  MUTATION_TYPE_UPDATE,
  MUTATION_TYPE_DELETE,
  CustomError,
  checkPermissionSimple,
  buildPermissionFilter,
} from 'shift-engine';

import {
  StorageDataTypeInteger,
  StorageDataTypeBigInt,
  StorageDataTypeNumeric,
  StorageDataTypeBoolean,
  StorageDataTypeText,
  StorageDataTypeJSON,
  StorageDataTypeTimestamp,
  StorageDataTypeTimestampTz,
  StorageDataTypeDate,
} from './storageDataTypes';

import Dataloader from 'dataloader';

const PERMISSION_TYPES = {
  read: 10,
  find: 20,
  mutation: 30,
}


const createDataLoader = (context, entity, modelRegistry, relayTypePromoterField) => {

  const loader = new Dataloader(ids => {

    const entityName = entity.name
    const { model, filterShaper } = modelRegistry[ entityName ]

    let permissionWhere = handlePermission(context, entity, PERMISSION_TYPES.read)
    if (permissionWhere) {
      permissionWhere = purifyFilter(
        filterShaper(permissionWhere)
      )
    }

    const baseWhere = {
      $and: []
    }

    if (ids.length === 1) {
      baseWhere.$and.push({
        id: ids[0]
      })
    }
    else {
      baseWhere.$and.push({
        id: {
          $in: ids,
        }
      })
    }

    const where = applyPermission(baseWhere, permissionWhere)

    return model.findAll({
      raw: true,
      where,
    })
    .then(data => {

      if (relayTypePromoterField) {
        data.map((row) => {
          row[ relayTypePromoterField ] = entityName
        })
      }

      return data
    })
    .then(data => {
      // reorder results to match requested IDs order
      // https://github.com/facebook/dataloader/issues/66
      return sortDataByKeys(ids, data)
    })
  })

  return loader
}


const processSequelizeError = (err, sequelize) => {

  if (err instanceof sequelize.UniqueConstraintError) {
    throw new CustomError('Uniqueness constraint violated', 'UniqueConstraintError')
  }
  else if (err instanceof sequelize.ForeignKeyConstraintError) {
    throw new CustomError('Foreign key not found', 'ForeignKeyConstraintError')
  }

  throw err
}


const buildInstanceNotFoundError = (id) => {
  return new CustomError(`Instance with ID '${id}' not found`, 'InstanceNotFound')
}


const buildInstanceNotFoundOrAccessDeniedError = (id) => {
  return new CustomError(`Instance with ID '${id}' not found or access denied`, 'InstanceNotFoundOrAccessDenied')
}


export const StorageTypePostgres = new StorageType({
  name: 'StorageTypePostgres',
  description: 'Postgres database storage type',

  findOne (entity, id, source, args, context, info, relayTypePromoterField) {

    const entityName = entity.name
    const loaders = context.loaders
    let loader = loaders[ entityName ]

    if (!loader) {
      const { modelRegistry } = this.getStorageInstanceFromContext(context)
      loader = loaders[ entityName ] = createDataLoader(context, entity, modelRegistry, relayTypePromoterField)
    }

    return loader.load(id)
  },


  async find (entity, source, args, context, info, relayTypePromoterField, parentConnection) {

    const {
      modelRegistry,
      // sequelize,
    } = this.getStorageInstanceFromContext(context)

    const entityName = entity.name
    const { model, filterShaper } = modelRegistry[ entityName ]

    const loaders = context.loaders
    let loader = loaders[ entityName ]

    const reverseOrder = !!args.last

    let permissionWhere = handlePermission(context, entity, PERMISSION_TYPES.find)
    if (permissionWhere) {
      permissionWhere = purifyFilter(
        filterShaper(permissionWhere)
      )
    }

    const order = args.orderBy.map(({attribute, direction}) => [
      attribute,
      (reverseOrder ? invertDirection(direction) : direction),
    ])

    let limit = getLimit(args.first, args.last)

    if (limit === 0) {
      // no point in fetching data
      return []
    }

    limit += 1 // ask for 1 more to determine if there is a next page


    const offset = args.offset

    const pageInfo = {
      hasNextPage: false,
      hasPreviousPage: false,
    }



    const baseWhere = {
      $and: [{
        ...processAndConvertFilter(entity, filterShaper, args, this),
        ...processCursors(entity, args)
      }]
    }

    const where = applyPermission(baseWhere, permissionWhere)


    if (parentConnection) {
      where.$and.push({
        [ parentConnection.attribute ]: parentConnection.id
      })
    }


    let result = await model.findAll({
      raw: true,
      limit,
      offset,
      order,
      where,
    })


    if (!loader) {
      loader = loaders[ entityName ] = createDataLoader(context, entity, modelRegistry, relayTypePromoterField)
    }

    result.map(row => {
      loader.prime(row.id, row)
    })


    // remove the 1 extra row if it exists
    if (result.length >= limit) {
      result = result.slice(0, limit -1)

      if (reverseOrder) {
        pageInfo.hasPreviousPage = true
      }
      else {
        pageInfo.hasNextPage = true
      }
    }

    return {
      data: reverseOrder ? result.reverse() : result,
      pageInfo,
    }
  },


  count (entity, source, args, context, info, parentConnection) {

    const { modelRegistry } = this.getStorageInstanceFromContext(context)

    const entityName = entity.name
    const { model, filterShaper } = modelRegistry[ entityName ]

    let permissionWhere = handlePermission(context, entity, PERMISSION_TYPES.find)
    if (permissionWhere) {
      permissionWhere = purifyFilter(
        filterShaper(permissionWhere)
      )
    }

    const baseWhere = {
      $and: []
    }

    const where = applyPermission(baseWhere, permissionWhere)

    if (parentConnection) {
      where.$and.push({
        [ parentConnection.attribute ]: parentConnection.id
      })
    }

    return model.count({
      where
    })

  },



  async mutate (entity, id, source, args, dataKey, entityMutation, context, info, relayTypePromoterField) {
    const entityName = entity.name
    const data = {
      ...args[ dataKey ]
    }
    const {
      modelRegistry,
      sequelize,
    } = this.getStorageInstanceFromContext(context)
    const { model, filterShaper } = modelRegistry[ entityName ]


    let permissionWhere = handlePermission(context, entity, PERMISSION_TYPES.mutation, entityMutation)
    if (permissionWhere) {
      permissionWhere = purifyFilter(
        filterShaper(permissionWhere)
      )
    }


    if (entityMutation.type === MUTATION_TYPE_CREATE) {
      let result

      try {
        result = await model.create(data)
      }
      catch (err) {
        processSequelizeError(err, sequelize)
      }

      result[ relayTypePromoterField ] = entityName

      return {
        [ dataKey ]: result,
      }
    }
    else if (entityMutation.type === MUTATION_TYPE_UPDATE) {

      let result
      const baseWhere = {
        $and: [{
          id
        }]
      }

      const where = applyPermission(baseWhere, permissionWhere)

      try {
        result = await model.update(data, {
          raw: true,
          returning: true,
          where
        })
      }
      catch (err) {
        processSequelizeError(err, sequelize)
      }

      const [ rowCount, returnedData ] = result

      if (rowCount < 1) {
        if (permissionWhere) {
          throw buildInstanceNotFoundOrAccessDeniedError(id)
        }

        throw buildInstanceNotFoundError(id)
      }

      returnedData[0][ relayTypePromoterField ] = entityName

      return {
        [ dataKey ]: returnedData[0],
      }
    }
    else if (entityMutation.type === MUTATION_TYPE_DELETE) {

      let rowCount
      const baseWhere = {
        $and: [{
          id
        }]
      }

      const where = applyPermission(baseWhere, permissionWhere)

      try {
        rowCount = await model.destroy({
          raw: true,
          returning: true,
          where,
        })
      }
      catch (err) {
        processSequelizeError(err, sequelize)
      }

      if (rowCount < 1) {
        if (permissionWhere) {
          throw buildInstanceNotFoundOrAccessDeniedError(id)
        }

        throw buildInstanceNotFoundError(id)
      }

      return {
        deleteRowCount: rowCount,
      }
    }

    return null

  }
})


export default StorageTypePostgres



StorageTypePostgres.addDataTypeMap(DataTypeID, StorageDataTypeBigInt);
StorageTypePostgres.addDataTypeMap(DataTypeInteger, StorageDataTypeInteger);
StorageTypePostgres.addDataTypeMap(DataTypeBigInt, StorageDataTypeBigInt);
StorageTypePostgres.addDataTypeMap(DataTypeFloat, StorageDataTypeNumeric);
StorageTypePostgres.addDataTypeMap(DataTypeBoolean, StorageDataTypeBoolean);
StorageTypePostgres.addDataTypeMap(DataTypeString, StorageDataTypeText);
StorageTypePostgres.addDataTypeMap(DataTypeJson, StorageDataTypeJSON);
StorageTypePostgres.addDataTypeMap(DataTypeTimestamp, StorageDataTypeTimestamp);
StorageTypePostgres.addDataTypeMap(DataTypeTimestampTz, StorageDataTypeTimestampTz);
StorageTypePostgres.addDataTypeMap(DataTypeDate, StorageDataTypeDate);



function invertDirection(direction) {
  return direction === 'DESC' ? 'ASC' : 'DESC'
}


function getLimit(first, last) {
  if (first >= 0) {
    return first
  }
  else if (last >= 0) {
    return last
  }

  return 10
}


function purifyFilter(filter) {
  const ret = {}

  Object.keys(filter).map(key => {
    if (typeof filter[ key ] !== 'undefined') {
      ret[ key ] = filter[ key ]
    }
  })

  return ret
}


function convertFilterLevel(filterShaper, filterLevel) {

  const converted = filterShaper(filterLevel)
  const filterLevelKeys = Object.keys(converted)
  const ret = {}

  filterLevelKeys.map(key => {
    const filter = filterLevel[ key ]

    if (filter) {
      if (isMap(filter)) {
        ret[ key ] = convertFilterLevel(filterShaper, filter)
      }
      else if (key === '$and' || key === '$or') {
        ret[ key ] = filter.map(item => convertFilterLevel(filterShaper, item))
      }
      else {
        ret[ key ] = converted[ key ]
      }
    }
  })


  filterLevelKeys.map(key => {
    if (typeof converted[ key ] !== 'undefined'  &&  typeof ret[ key ] === 'undefined') {
      ret[ key ] = converted[ key ]
    }
  })

  return ret
}


function processAndConvertFilter(entity, filterShaper, args) {
  const where = processFilter(entity, args, StorageTypePostgres)
  const convertedWhere = convertFilterLevel(filterShaper, where)

  return convertedWhere
}



function applyPermission(where, permissionWhere) {

  if (!permissionWhere) {
    return where
  }

  const newWhere = { ...where }
  newWhere.$and.push(permissionWhere)

  return newWhere
}


function loadPermission (entity, permissionType, entityMutation) {

  if (!entity.permissions) {
    return null
  }
  else if (permissionType === PERMISSION_TYPES.read) {
    return entity.permissions.read
  }
  else if (permissionType === PERMISSION_TYPES.find) {
    return entity.permissions.find
  }
  else if (permissionType === PERMISSION_TYPES.mutation) {
    const mutationName = entityMutation.name
    return entity.permissions.mutations[ mutationName ]
  }

  throw new CustomError(`Unknown permission type '${permissionType}'`, 'PermissionTypeError')
}


function handlePermission(context, entity, permissionType, entityMutation) {

  const permission = loadPermission(entity, permissionType, entityMutation)

  if (!permission) {
    return null
  }

  const {
    user: {
      id: userId,
      roles: userRoles,
    }
  } = context.req

  const permitted = checkPermissionSimple(permission, userId, userRoles)

  if (permitted) {
    return null
  }

  const permissionFilter = buildPermissionFilter(permission, userId)

  if (!permissionFilter) {
    throw new CustomError('Access denied', 'PermissionError')
  }

  return permissionFilter
}
