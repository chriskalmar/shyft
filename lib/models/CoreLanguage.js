
import {
  dataTypes,
  Entity,
} from '../../';


const {
  DataTypeString,
} = dataTypes



export const CoreLanguage = new Entity({
  name: 'language',
  domain: 'core',
  description: 'A language',

  // indexing: {
  //   unique: [
  //     [ 'name' ],
  //     [ 'iso_code' ],

  attributes: {

    name: {
      type: DataTypeString,
      description: 'The name of the language',
      maxLength: 30,
      required: true
    },

    nativeName: {
      type: DataTypeString,
      description: 'The native name of the language',
      maxLength: 30,
      required: true
    },

    isoCode: {
      type: DataTypeString,
      description: 'ISO code of the language',
      pattern: '^[a-z]+$',
      minLength: 2,
      maxLength: 2,
      required: true
    },
  }

})

