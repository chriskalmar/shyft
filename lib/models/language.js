
'use script';

module.exports = {

  domain: 'core',

  entities: [

    {
      name: 'language',
      description: 'a language',

      cascadeOnDelete: true,

      attributes: [

        {
          name: 'name',
          type: 'string',
          description: 'Name',
          maxLength: 30,
          required: true
        },

        {
          name: 'native_name',
          type: 'string',
          description: 'Name',
          maxLength: 30,
          required: true
        },

        {
          name: 'iso_code',
          type: 'string',
          description: 'ISO 639-1 code of the language',
          maxLength: 2,
          required: true
        },

      ],

      indexing: {
        unique: [
          [ 'name' ],
          [ 'iso_code' ],
        ]
      }

    }

  ]

}
