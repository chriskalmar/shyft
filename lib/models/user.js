
'use script';

module.exports = {

  domain: 'core',

  entities: [

    {
      name: 'user',
      description: 'a user',

      isUserTable: true,
      cascadeOnDelete: true,

      attributes: [

        {
          name: 'username',
          type: 'string',
          description: 'Username',
          pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
          minLength: 3,
          maxLength: 30,
          required: true
        },

        {
          name: 'first_name',
          type: 'string',
          description: 'First name',
          maxLength: 30
        },

        {
          name: 'last_name',
          type: 'string',
          description: 'Last name',
          maxLength: 30
        },

        {
          name: 'language',
          type: 'reference',
          description: 'User language',
          target: 'core::language',
          required: true
        },

        {
          name: 'email',
          type: 'email',
          description: 'Email address',
          required: true
        },

        {
          name: 'password',
          type: 'password',
          description: 'User password',
          required: true
        }
      ],

      indexing: {
        unique: [
          [ 'username' ],
          [ 'email' ],
        ]
      }

    }

  ]

}
