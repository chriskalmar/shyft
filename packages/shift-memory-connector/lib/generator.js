
import { isEntity } from 'shift-engine';
import StorageTypeMemory from './StorageTypeMemory';
import _ from 'lodash';



export const generateMemoryDB = (schema) => {

  const memoryDB = {}

  _.map(schema.getEntities(), (entity, entityName) => {

    if (String(entity.storageType) !== String(StorageTypeMemory)) {
      return
    }

    const model = {
      name: entityName,
      description: entity.description,

      fields: () => {
        const fields = {}

        _.map(entity.getAttributes(), (attribute) => {

          // skip for computed values
          if (attribute.resolve) {
            return
          }

          memoryDB[ entityName ].fieldTransformerMap[ attribute.gqlFieldName ] = attribute.name

          const field = {
            description: attribute.description,
          };


          // it's a reference
          if (isEntity(attribute.type)) {

            const targetEntity = attribute.type
            const targetEntityName = targetEntity.name

            field.type = memoryDB[ targetEntityName ].model
          }
          // it's a regular attribute
          else {
            field.type = attribute.type
          }

          fields[ attribute.name ] = field;

        });

        return fields
      }
    }



    memoryDB[ entityName ] = {
      data: [],
      model,
      fieldTransformerMap: {},
    }

  })



  _.map(memoryDB, (entity) => {

    // lazy generate fields
    entity.model.fields = entity.model.fields()

  })



  return memoryDB;

}
