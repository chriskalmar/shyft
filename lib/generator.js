
import StorageTypePostgres from './StorageTypePostgres';
import _ from 'lodash';
import toposort from 'toposort';
import { generateIndexName } from './util';


import {
  isEntity,
  INDEX_UNIQUE,
  INDEX_GENERIC,
  mapOverProperties,
  isDataTypeEnum,
} from 'shift-engine';

import {
  shaper,
} from 'json-shaper'


const filterOperatorMap = {
  $ne: '$ne',

  $in: '$in',
  $notIn: '$not_in',

  $lt: '$lt',
  $lte: '$lte',

  $gt: '$gt',
  $gte: '$gte',

  $startsWith: '$starts_with',
  $contains: '$contains',
  $endsWith: '$ends_with',

  $notStartsWith: '$not_starts_with',
  $notContains: '$not_contains',
  $notEndsWith: '$not_ends_with',

  $not: '$not',
  $or: '$or',
  $and: '$and',
}


export const loadModels = (schema) => {

  const modelRegistry = {}

  _.map(schema.getEntities(), (entity, entityName) => {

    if (String(entity.storageType) !== String(StorageTypePostgres)) {
      return
    }

    const attributes = {}
    const references = []
    const dataShaperMap = {}
    const filterShaperMap = {}

    _.map(entity.getAttributes(), (attribute) => {

      // skip for computed values
      if (attribute.resolve) {
        return
      }

      const attributeName = attribute.name
      const storageAttributeName = attribute.meta && attribute.meta.storageAttributeName
        ? attribute.meta.storageAttributeName
        : _.snakeCase(attribute.name)

      dataShaperMap[ attributeName ] = storageAttributeName
      filterShaperMap[ attributeName ] = attributeName

      let storageDataType

      // it's a reference
      if (isEntity(attribute.type)) {
        const primaryAttribute = attribute.type.getPrimaryAttribute()
        storageDataType = StorageTypePostgres.convertToStorageDataType(primaryAttribute.type)

        if (attribute.targetAttributesMap) {
          // sequelize doesn't support composite foreign keys yet (so this has no impact):
          // https://github.com/sequelize/sequelize/issues/311
          mapOverProperties(attribute.targetAttributesMap, (targetAttribute, sourceAttributeName) => {
            references.push({
              sourceAttributeName,
              targetAttributeName: targetAttribute.name,
              targetEntityName: attribute.type.name,
            })
          })
        }
        else {
          references.push({
            sourceAttributeName: attributeName,
            targetAttributeName: primaryAttribute.name,
            targetEntityName: attribute.type.name,
          })
        }
      }
      // it's a regular attribute
      else {
        storageDataType = StorageTypePostgres.convertToStorageDataType(attribute.type)
      }


      if (isDataTypeEnum(attribute.type)) {
        throw new Error('data type `enum` is not yet supported')
        // attributes[ attributeName ] = {
        //   type: storageDataType.nativeDataType,
        //   enum: attribute.type.values,
        //   nullable: !attribute.required,
        // }
      }
      else {
        attributes[ attributeName ] = {
          name: storageAttributeName,
          type: storageDataType.nativeDataType,
          primary: attribute.isPrimary,
          nullable: !attribute.required,
        }

        if (attribute.isPrimary) {
          attributes[attributeName].generated = (storageDataType.name === 'StorageDataTypeUUID')
            ? 'uuid'
            : 'increment'
        }

      }

    })



    const indices = {}

    if (entity.indexes) {

      entity.indexes.map((index) => {

        if (index.type === INDEX_UNIQUE || index.type === INDEX_GENERIC) {

          const indexAttributes = index.attributes.map(_.snakeCase)
          const indexName = generateIndexName(_.snakeCase(entityName), indexAttributes, 'key')

          indices[indexName]={
            target: entityName,
            columns: index.attributes,
            unique: index.type === INDEX_UNIQUE,
          }
        }

      })

    }

    const model = {
      name: entityName,
      columns: attributes,
      table: {
        name: _.snakeCase(entityName),
      },
      indices
    }

    modelRegistry[ entityName ] = {
      model,
      dataShaperMap,
      dataShaper: shaper(dataShaperMap),
      reverseDataShaper: shaper(_.invert(dataShaperMap)),
      filterShaper: shaper(_.extend(filterShaperMap, filterOperatorMap)),
      references,
    }

  })



  _.map(modelRegistry, (entity) => {
    entity.references.map(({sourceAttributeName, targetEntityName}) => {

      entity.model.relations = entity.model.relations || {}
      entity.model.relations[sourceAttributeName] = {
        target: targetEntityName,
        type: 'many-to-one',
        joinColumn: {
          name: entity.model.columns[sourceAttributeName].name,
        },
        onDelete: 'CASCADE'
      }

    })
  })


  return modelRegistry
}



export const generateMockData = async (storageInstance, schema) => {

  const edges = []
  const entities = schema.getEntities()
  const mockInstancesMemory = {}

  _.map(entities, (entity) => {
    _.map(entity.getAttributes(), (attribute) => {
      if (isEntity(attribute.type)) {

        // skip cyclic dependencies
        if (entity.name === attribute.type.name) {
          return
        }

        edges.push([
          attribute.type.name,
          entity.name
        ])
      }
    })
  })

  const dependencySortedEntityNames = toposort(edges)

  // add independent entities
  _.map(entities, (entity) => {
    if (dependencySortedEntityNames.indexOf(entity.name) === -1) {
      dependencySortedEntityNames.push(entity.name)
    }
  })

  for (let e=0; e < dependencySortedEntityNames.length; e++) {
    const entityName = dependencySortedEntityNames[ e ]
    const entity = entities[ entityName ]

    if (String(entity.storageType) !== String(StorageTypePostgres)) {
      return
    }

    try {
      for(let i=0; i < _.random(10, 100); i++) {
        await generateItem(storageInstance, entity, mockInstancesMemory)
      }
    }
    catch(err) {
      console.error(err)
    }

  }

}



async function generateItem(storageInstance, entity, mockInstancesMemory) {

  const entityName = entity.name
  const item = {}

  _.map(entity.getAttributes(), ({ type, isPrimary, required, mock }, name) => {

    if (isPrimary) {
      return
    }

    if (!required && Math.random() > 0.5) {
      return
    }

    if (isEntity(type)) {
      item[ name ] = _.sample(mockInstancesMemory[ type.name ])
    }
    else {
      const mockDataGenerator = mock || type.mock

      if (mockDataGenerator) {
        item[ name ] = mockDataGenerator()
      }
    }
  })


  const manager = storageInstance.manager;
  let instance

  const result = await manager
    .createQueryBuilder()
    .insert()
    .into(entityName)
    .values(item)
    .returning('*')
    .execute()

  if (result.length) {
    instance = result[0]

    const { name: primaryAttributeName } = entity.getPrimaryAttribute()

    mockInstancesMemory[ entityName ] = mockInstancesMemory[ entityName ] || []
    mockInstancesMemory[entityName].push(instance[primaryAttributeName])
  }

}
