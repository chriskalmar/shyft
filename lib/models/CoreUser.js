
import {
  datatypes,
  Entity,
} from '../../';

import CoreLanguage from './CoreLanguage';

const {
  DataTypeString,
} = datatypes

export const CoreUser = new Entity({
  name: 'user',
  domain: 'core',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  // indexing: {
  //   unique: [
  //     [ 'username' ],
  //     [ 'email' ],


  attributes: {

    username: {
      type: DataTypeString,
      description: 'Username',
      pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
      minLength: 3,
      maxLength: 30,
      required: true
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
      maxLength: 30
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
      maxLength: 30
    },

    language: {
      type: CoreLanguage,
      description: 'User language',
      required: true
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true
    }
  }

})



