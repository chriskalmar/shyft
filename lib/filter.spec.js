import { createConnection } from 'typeorm';
import {
  buildWhereQuery,
  purifyFilter,
} from './filter';
import { StorageTypePostgres } from './StorageTypePostgres';
import { loadModels } from './generator';
import { Server } from '../test/models/Server';

import {
  Schema,
} from 'shift-engine';


let connection

beforeAll(async () => {

  const schema = new Schema({

    defaultStorageType: StorageTypePostgres,

    entities: [
      Server,
    ],
  })


  const modelRegistry = loadModels(schema)

  const entitySchemas = Object.keys(modelRegistry).map(entityName => {
    return modelRegistry[entityName].model
  })


  connection = await createConnection({
    database: ':memory:',
    type: 'sqlite',
    entitySchemas,
  });
})



describe('purifyFilter', () => {

  it('should clean empty filter bocks', () => {
    const filter = {
      $and: [
        {
          username: {
            $contains: 'bot',
            $startsWith: '_system_%'
          },
          language: {
            $in: [],
            $lt: 5,
            $gte: 1,
          },
          $or: [
            {
              'LookupEntity.role': 1
            },
            {
            },
            {
              accessLevel: 1
            }
          ]
        },
        {
          $and: [
            {},
            {}
          ]
        },
        {
          $or: [
            {},
            {
              $and: [
                {
                  username: 'hallo'
                }
              ]
            }
          ]
        }
      ],
      $or: [
        {
          lastname: 'smith',
          firtname: {}
        },
        {
        }
      ]
    }

    const pureFilter = purifyFilter(filter)
    expect(pureFilter).toMatchSnapshot()
  })
})


describe('filter', () => {

  describe('buildWhereQuery', () => {

    const entityNameServer = 'Server'

    it('empty filter', () => {
      const qBuilder = connection.createQueryBuilder(entityNameServer, entityNameServer)
      const filter = null
      const pureFilter = purifyFilter(filter)

      buildWhereQuery(qBuilder, pureFilter, entityNameServer)
      const query = qBuilder.getQueryAndParameters()
      expect(query).toMatchSnapshot()
    })


    it('simple properties filter', () => {
      const qBuilder = connection.createQueryBuilder(entityNameServer, entityNameServer)
      const filter = {
        ip: '127.0.0.1',
        name: 'test-server-1',
      }
      const pureFilter = purifyFilter(filter)

      buildWhereQuery(qBuilder, pureFilter, entityNameServer)
      const query = qBuilder.getQueryAndParameters()
      expect(query).toMatchSnapshot()
    })


    it('logical AND filter', () => {
      const qBuilder = connection.createQueryBuilder(entityNameServer, entityNameServer)
      const filter = {
        '$and': [
          {
            ip: '127.0.0.1',
            name: 'test-server-1',
          },
          {
            clusterZone: 4,
          }
        ]
      }
      const pureFilter = purifyFilter(filter)

      buildWhereQuery(qBuilder, pureFilter, entityNameServer)
      const query = qBuilder.getQueryAndParameters()
      expect(query).toMatchSnapshot()
    })


    it('logical OR filter', () => {
      const qBuilder = connection.createQueryBuilder(entityNameServer, entityNameServer)
      const filter = {
        '$or': [
          {
            ip: '127.0.0.1',
            name: 'test-server-1',
          },
          {
            ip: '127.0.0.1',
            name: 'test-server-2',
          }
        ]
      }
      const pureFilter = purifyFilter(filter)

      buildWhereQuery(qBuilder, pureFilter, entityNameServer)
      const query = qBuilder.getQueryAndParameters()
      expect(query).toMatchSnapshot()
    })


    it('combined logical AND/OR filter', () => {
      const qBuilder1 = connection.createQueryBuilder(entityNameServer, entityNameServer)

      const filter1 = {
        '$and': [
          {
            ip: '127.0.0.1',
            $or: [
              {
                clusterZone: 1,
              },
              {
              },
              {
                clusterZone: 7,
              }
            ]
          },
          {
            name: 'test-server-1',
          }
        ]
      }
      const pureFilter1 = purifyFilter(filter1)

      buildWhereQuery(qBuilder1, pureFilter1, entityNameServer)
      const query1 = qBuilder1.getQueryAndParameters()
      expect(query1).toMatchSnapshot()


      const qBuilder2 = connection.createQueryBuilder(entityNameServer, entityNameServer)

      const filter2 = {
        '$or': [
          {
            ip: '127.0.0.1',
            $and: [
              {
                clusterZone: 1,
              },
              {
              },
              {
                name: 'test-server-110',
              }
            ]
          },
          {
            name: 'test-server-1',
          }
        ]
      }
      const pureFilter2 = purifyFilter(filter2)

      buildWhereQuery(qBuilder2, pureFilter2, entityNameServer)
      const query2 = qBuilder2.getQueryAndParameters()
      expect(query2).toMatchSnapshot()
    })
  })

})
