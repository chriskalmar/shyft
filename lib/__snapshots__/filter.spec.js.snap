// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter buildWhereQuery combined logical AND/OR filter 1`] = `
Array [
  "SELECT \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\", \\"Server\\".\\"id\\" AS \\"Server_id\\" FROM \\"server\\" \\"Server\\" WHERE ((\\"Server\\".\\"ip\\" = $1 AND ((\\"Server\\".\\"cluster_zone\\" = $2) OR (\\"Server\\".\\"cluster_zone\\" = $3))) AND (\\"Server\\".\\"name\\" = $4))",
  Array [
    "127.0.0.1",
    1,
    7,
    "test-server-1",
  ],
]
`;

exports[`filter buildWhereQuery combined logical AND/OR filter 2`] = `
Array [
  "SELECT \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\", \\"Server\\".\\"id\\" AS \\"Server_id\\" FROM \\"server\\" \\"Server\\" WHERE (((\\"Server\\".\\"ip\\" = $1 AND (\\"Server\\".\\"cluster_zone\\" = $2) AND (\\"Server\\".\\"name\\" = $3)) OR (\\"Server\\".\\"name\\" = $4)))",
  Array [
    "127.0.0.1",
    1,
    "test-server-110",
    "test-server-1",
  ],
]
`;

exports[`filter buildWhereQuery empty filter 1`] = `
Array [
  "SELECT \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\", \\"Server\\".\\"id\\" AS \\"Server_id\\" FROM \\"server\\" \\"Server\\"",
  Array [],
]
`;

exports[`filter buildWhereQuery logical AND filter 1`] = `
Array [
  "SELECT \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\", \\"Server\\".\\"id\\" AS \\"Server_id\\" FROM \\"server\\" \\"Server\\" WHERE ((\\"Server\\".\\"ip\\" = $1 AND \\"Server\\".\\"name\\" = $2) AND (\\"Server\\".\\"cluster_zone\\" = $3))",
  Array [
    "127.0.0.1",
    "test-server-1",
    4,
  ],
]
`;

exports[`filter buildWhereQuery logical OR filter 1`] = `
Array [
  "SELECT \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\", \\"Server\\".\\"id\\" AS \\"Server_id\\" FROM \\"server\\" \\"Server\\" WHERE (((\\"Server\\".\\"ip\\" = $1 AND \\"Server\\".\\"name\\" = $2) OR (\\"Server\\".\\"ip\\" = $3 AND \\"Server\\".\\"name\\" = $4)))",
  Array [
    "127.0.0.1",
    "test-server-1",
    "127.0.0.1",
    "test-server-2",
  ],
]
`;

exports[`filter buildWhereQuery reject if AND is not provided with an array 1`] = `"$and expects an array of filters"`;

exports[`filter buildWhereQuery reject if OR is not provided with an array 1`] = `"$or expects an array of filters"`;

exports[`filter buildWhereQuery reject if unknown operators are used 1`] = `"Unknown filter operator: $unknownOp"`;

exports[`filter buildWhereQuery simple properties filter 1`] = `
Array [
  "SELECT \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\", \\"Server\\".\\"id\\" AS \\"Server_id\\" FROM \\"server\\" \\"Server\\" WHERE (\\"Server\\".\\"ip\\" = $1 AND \\"Server\\".\\"name\\" = $2)",
  Array [
    "127.0.0.1",
    "test-server-1",
  ],
]
`;

exports[`filter purifyFilter should clean empty filter bocks 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "LookupEntity.role": 1,
        },
        Object {
          "accessLevel": 1,
        },
      ],
      "language": Object {
        "$gte": 1,
        "$lt": 5,
      },
      "username": Object {
        "$contains": "bot",
        "$startsWith": "_system_%",
      },
    },
    Object {
      "$or": Array [
        Object {
          "$and": Array [
            Object {
              "username": "hallo",
            },
          ],
        },
      ],
    },
  ],
  "$or": Array [
    Object {
      "lastname": "smith",
    },
  ],
}
`;
